{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "import ete3\n",
    "import Bio.AlignIO\n",
    "import Bio.Seq\n",
    "import Bio.Align\n",
    "import numpy\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get amino acid alignment\n",
    "aln_cds = Bio.AlignIO.read('../../0_data/selectionAndASR/5133_NT_AL.fasta', format='fasta')\n",
    "print(aln_cds)\n",
    "with open('5133_NT_AL.aa.fasta', 'w') as f:\n",
    "    for seq_record in aln_cds:\n",
    "        print(f'>{seq_record.id}', file=f)\n",
    "        print(f'{seq_record.seq.translate()}', file=f)\n",
    "aln_aa = Bio.AlignIO.read('5133_NT_AL.aa.fasta', format='fasta')\n",
    "print(aln_aa)\n",
    "\n",
    "sp_list = [x.id for x in aln_aa]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import tree\n",
    "tree = ete3.Tree('../../0_data/selectionAndASR/timetree_all_sp.treefile')\n",
    "tree_sps = tree.get_leaf_names()\n",
    "\n",
    "# remove gap-rich sequences\n",
    "out_dir = '../03_orthomam_filterseq10/'\n",
    "os.makedirs(out_dir, exist_ok=True)\n",
    "# aln_cds = Bio.AlignIO.read('5133_NT_AL.fasta', format='fasta')\n",
    "total_length = aln_cds.get_alignment_length()\n",
    "filtered_sp_list = []\n",
    "filtered_seqrec_list = []\n",
    "for seq_record in aln_cds:\n",
    "    if str(seq_record.seq).count('-') > 0.1 * total_length:\n",
    "        continue\n",
    "    else:\n",
    "        filtered_sp_list.append(seq_record.id)\n",
    "        filtered_seqrec_list.append(seq_record)\n",
    "filtered_aln_cds = Bio.Align.MultipleSeqAlignment(filtered_seqrec_list)\n",
    "filtered_aln_cds_arr = numpy.array(filtered_aln_cds)\n",
    "nongap_indices = [x for x in range(filtered_aln_cds_arr.shape[1])\n",
    "                  if not numpy.all(filtered_aln_cds_arr[:, x] == '-')]\n",
    "filtered_aln_cds_arr = filtered_aln_cds_arr[:, nongap_indices]\n",
    "for i in range(len(filtered_aln_cds)):\n",
    "    filtered_aln_cds[i].seq = Bio.Seq.Seq(''.join(filtered_aln_cds_arr[i]))\n",
    "print(filtered_aln_cds)\n",
    "need_sps = list(set(filtered_sp_list)&set(tree_sps))\n",
    "with open(f'{out_dir}/5133_NT_AL.filtered10.fasta', 'w') as f:\n",
    "    for seq_record in filtered_aln_cds:\n",
    "        if seq_record.id in need_sps:\n",
    "            print(f'>{seq_record.id}', file=f)\n",
    "            print(f'{seq_record.seq}', file=f)\n",
    "\n",
    "with open(f'{out_dir}/5133_NT_AL.filtered10.aa.fasta', 'w') as f:\n",
    "    for seq_record in filtered_aln_cds:\n",
    "        if seq_record.id in need_sps:\n",
    "            print(f'>{seq_record.id}', file=f)\n",
    "            print(f'{seq_record.seq.translate()}', file=f)\n",
    "aln_aa = Bio.AlignIO.read(f'{out_dir}5133_NT_AL.filtered10.aa.fasta', format='fasta')\n",
    "\n",
    "# get filtered species tree\n",
    "tree.prune(need_sps, preserve_branch_length=True)\n",
    "print(len(tree.get_leaf_names()))\n",
    "with open(f'{out_dir}/timetree.treefile', 'w') as f:\n",
    "    print(tree.write(format=9), file=f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make selection results dictories\n",
    "def mkdir(path):\n",
    "\tfolder = os.path.exists(path)\n",
    "\tif not folder:                  \n",
    "\t\tos.makedirs(path, exist_ok=True)            \n",
    "\t\t\n",
    "file = f'{out_dir}/' \n",
    "splist = ['primate','rodent'] \n",
    "dirlist = ['busted',  'relax', 'busted_within_clade', 'relax_within_clade'] \n",
    "for sp in splist:\n",
    "    for dir in dirlist:\n",
    "        mkdir(os.path.join(file, sp, dir))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# build selection tree__primate\n",
    "out_dir_s = f'{out_dir}/rodent'\n",
    "\n",
    "# MRCA\n",
    "tree = ete3.Tree(f'{out_dir}/timetree.treefile')\n",
    "selection_node = tree.get_common_ancestor([tree & 'Homo_sapiens', tree & 'Lemur_catta'])\n",
    "selection_node.name = str(selection_node.name) + '{Foreground}'\n",
    "with open(f'{out_dir_s}/busted/timetree_filtered10_fg.nw', 'w') as f:\n",
    "    print(tree.write(format=1), file=f)\n",
    "tree.write(format=1, outfile=f'{out_dir_s}/relax/timetree_filtered10_fg.nw')\n",
    "\n",
    "# within_clade\n",
    "tree = ete3.Tree(f'{out_dir}/timetree.treefile')\n",
    "selection_node = tree.get_common_ancestor([tree & 'Homo_sapiens', tree & 'Lemur_catta'])\n",
    "for node in selection_node.iter_descendants(\"postorder\"):\n",
    "    node.name = str(node.name)+'{Foreground}'\n",
    "tree.write(format=1, outfile=f'{out_dir_s}/relax_within_clade/timetree_filtered10_fg_inclade_nosp.nw')\n",
    "tree.write(format=1, outfile=f'{out_dir_s}/busted_within_clade/timetree_filtered10_fg_inclade_nosp.nw')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# build selection tree__rodent\n",
    "out_dir_s = f'{out_dir}/rodent'\n",
    "\n",
    "# MRCA\n",
    "tree = ete3.Tree(f'{out_dir}/timetree.treefile')\n",
    "selection_node = tree.get_common_ancestor([tree & 'Mus_musculus', tree & 'Marmota_monax'])\n",
    "selection_node.name = str(selection_node.name) + '{Foreground}'\n",
    "with open(f'{out_dir_s}/busted/timetree_filtered10_fg.nw', 'w') as f:\n",
    "    print(tree.write(format=1), file=f)\n",
    "tree.write(format=1, outfile=f'{out_dir_s}/relax/timetree_filtered10_fg.nw')\n",
    "\n",
    "# within_clade\n",
    "tree = ete3.Tree(f'{out_dir}/timetree.treefile')\n",
    "selection_node = tree.get_common_ancestor([tree & 'Mus_musculus', tree & 'Marmota_monax'])\n",
    "for node in selection_node.iter_descendants(\"postorder\"):\n",
    "    node.name = str(node.name)+'{Foreground}'\n",
    "tree.write(format=1, outfile=f'{out_dir_s}/relax_within_clade/timetree_filtered10_fg_inclade_nosp.nw')\n",
    "tree.write(format=1, outfile=f'{out_dir_s}/busted_within_clade/timetree_filtered10_fg_inclade_nosp.nw')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "dirs = []\n",
    "for dir in os.listdir(out_dir):\n",
    "    if dir.endswith(\"10.fasta\"):\n",
    "        fa_file = os.path.join(out_dir, dir)\n",
    "    if \".\" not in dir:\n",
    "        dirs.append(dir)\n",
    "\n",
    "for dir in dirs:\n",
    "    for root, dirss, files in os.walk(os.path.join(out_dir, dir)):\n",
    "        for subdir in dirss:\n",
    "            json_files = [file for file in os.listdir(os.path.join(root, subdir)) if file.endswith('.json')]\n",
    "            if not json_files:\n",
    "                tree_file = [file for file in os.listdir(os.path.join(root, subdir)) if file.endswith('.nw')]\n",
    "                tree_file = os.path.join(root, subdir, tree_file[0])\n",
    "                if 'busted' in subdir:\n",
    "                    command = f'hyphy CPU=1 busted --alignment {fa_file} --branches Foreground --tree {tree_file} --output {root}/{subdir}/{subdir}_{fa_file.split(\"/\")[-1].replace(\".fasta\",\".json\")} > {root}/{subdir}/{subdir}_{fa_file.split(\"/\")[-1].replace(\".fasta\",\".out\")} &'\n",
    "                elif 'relax' in subdir:\n",
    "                    command = f'hyphy CPU=1 relax --alignment {fa_file} --test Foreground --tree {tree_file} --output {root}/{subdir}/{subdir}_{fa_file.split(\"/\")[-1].replace(\".fasta\",\".json\")} > {root}/{subdir}/{subdir}_{fa_file.split(\"/\")[-1].replace(\".fasta\",\".out\")} &'\n",
    "                else:\n",
    "                    command = ''\n",
    "                os.system(command)\n",
    "                # print(command)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "glycosylation",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
